test/built-ins/RegExp/named-groups/groups-object-undefined.js

```diff
@@ -0,0 +1,99 @@
+includes: [propertyHelper.js]
+esid: sec-regexpbuiltinexec
+features: [regexp-named-groups]
+info: >
+  Runtime Semantics: RegExpBuiltinExec ( R, S )
+    24. If _R_ contains any |GroupName|, then
+      a. Let _groups_ be ObjectCreate(*null*).
+    25. Else,
+      a. Let _groups_ be *undefined*.
+    26. Perform ! CreateDataProperty(_A_, `"groups"`, _groups_).
+---*/
+
+// https://github.com/tc39/proposal-regexp-named-groups/pull/40
+{
+  const re = /./;
+  const result = re.exec("a");
+  assert.sameValue(result.__proto__, Array.prototype);
+  assert(result.hasOwnProperty("groups"));
+  assert.sameValue("a", result[0]);
+  assert.sameValue(0, result.index);
+  assert.sameValue(undefined, result.groups);
+  verifyProperty(result, "groups", {
+    writable: true,
+    enumerable: true,
+    configurable: true,
+  });
+
+  Array.prototype.groups = { a: "b" };
+  assert.sameValue("$<a>", "a".replace(re, "$<a>"));
+  Array.prototype.groups = undefined;
+}
+
+{
+  const re = /(?<a>a).|(?<x>x)/;
+  const result = re.exec("ab");
+  assert.sameValue(result.__proto__, Array.prototype);
+  assert(result.hasOwnProperty("groups"));
+  assert.sameValue("ab", result[0]);
+  assert.sameValue("a", result[1]);
+  assert.sameValue(undefined, result[2]);
+  assert.sameValue(0, result.index);
+  assert.sameValue("a", result.groups.a);
+  assert.sameValue(undefined, result.groups.x);
+
+  // `a` is a matched named capture, `b` is an unmatched named capture, and `z`
+  // is not a named capture.
+  Array.prototype.groups = { a: "b", x: "y", z: "z" };
+  assert.sameValue("a", "ab".replace(re, "$<a>"));
+  assert.sameValue("", "ab".replace(re, "$<x>"));
+  assert.sameValue("", "ab".replace(re, "$<z>"));
+  Array.prototype.groups = undefined;
+}
+
+{
+  class FakeRegExp extends RegExp {
+    exec(subject) {
+      const fakeResult = ["ab", "a"];
+      fakeResult.index = 0;
+      // `groups` is not set, triggering prototype lookup.
+      return fakeResult;
+    }
+  };
+
+  const re = new FakeRegExp();
+  const result = re.exec("ab");
+  assert.sameValue(result.__proto__, Array.prototype);
+  assert.sameValue(false, result.hasOwnProperty("groups"));
+
+  Array.prototype.groups = { a: "b" };
+  Array.prototype.groups.__proto__.b = "c";
+  assert.sameValue("b", "ab".replace(re, "$<a>"));
+  assert.sameValue("c", "ab".replace(re, "$<b>"));
+  Array.prototype.groups = undefined;
+}
+
+{
+  class FakeRegExp extends RegExp {
+    exec(subject) {
+      const fakeResult = ["ab", "a"];
+      fakeResult.index = 0;
+      fakeResult.groups = { a: "b" };
+      fakeResult.groups.__proto__.b = "c";
+      return fakeResult;
+    }
+  };
+
+  const re = new FakeRegExp();
+  const result = re.exec("ab");
+  assert.sameValue(result.__proto__, Array.prototype);
+  assert(result.hasOwnProperty("groups"));
+  assert.sameValue("b", result.groups.a);
+  assert.sameValue("b", "ab".replace(re, "$<a>"));
+  assert.sameValue("c", "ab".replace(re, "$<b>"));
+}
```

Comment:
Each of these blocks could be a different file. This is not a blocker, and I will likely do it myself after I land this, but I figure I should give you the opportunity to make the split the way you want to. Let me know what you want to do.

Reply:
@rwaldron I’ll do the split — no worries! Naming things is hard, though… Gotta think of a good file name :)